src/ILCompiler.Compiler/src/Compiler/RyuJitCompilation.cs

```diff
@@ -0,0 +1,141 @@
+using ILCompiler.DependencyAnalysisFramework;
+
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.JitInterface;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    public sealed class RyuJitCompilation : Compilation
+    {
+        private CorInfoImpl _corInfo;
+        private JitConfigProvider _jitConfigProvider;
+
+        internal RyuJitCompilation(
+            DependencyAnalyzerBase<NodeFactory> dependencyGraph,
+            NodeFactory nodeFactory,
+            Logger logger,
+            JitConfigProvider configProvider)
+            : base(dependencyGraph, nodeFactory, new NameMangler(false), logger)
+        {
+            _jitConfigProvider = configProvider;
+        }
+
+        protected override void CompileInternal(string outputFile)
+        {
+            _corInfo = new CorInfoImpl(this, _jitConfigProvider);
+
+            var nodes = _dependencyGraph.MarkedNodeList;
+
+            ObjectWriter.EmitObject(outputFile, nodes, NodeFactory);
+        }
+
+        protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCore<NodeFactory>> obj)
+        {
+            foreach (DependencyNodeCore<NodeFactory> dependency in obj)
+            {
+                var methodCodeNodeNeedingCode = dependency as MethodCodeNode;
+                if (methodCodeNodeNeedingCode == null)
+                {
+                    // To compute dependencies of the shadow method that tracks dictionary
+                    // dependencies we need to ensure there is code for the canonical method body.
+                    var dependencyMethod = (ShadowConcreteMethodNode<MethodCodeNode>)dependency;
+                    methodCodeNodeNeedingCode = dependencyMethod.CanonicalMethodNode;
+                }
+
+                // We might have already compiled this method.
+                if (methodCodeNodeNeedingCode.StaticDependenciesAreComputed)
+                    continue;
+
+                MethodDesc method = methodCodeNodeNeedingCode.Method;
+
+                if (Logger.IsVerbose)
+                {
+                    string methodName = method.ToString();
+                    Logger.Writer.WriteLine("Compiling " + methodName);
+                }
+
+                try
+                {
+                    _corInfo.CompileMethod(methodCodeNodeNeedingCode);
+                }
+                catch (TypeSystemException ex)
+                {
+                    // TODO: fail compilation if a switch was passed
+                    if (!TryCompileWithThrowingBody(methodCodeNodeNeedingCode, ex))
+                        throw;
+                    // TODO: Log as a warning
+                }
+            }
+        }
+
+        /// <summary>
+        /// Compiles the provided method code node while swapping its body with a throwing stub.
+        /// </summary>
+        private bool TryCompileWithThrowingBody(MethodCodeNode methodNode, TypeSystemException exception)
+        {
```

Comment:
I looked into this, but this is same argument as above: this depends on `MethodCodeNode` that is RyuJIT specific. Factoring it out will make us share more code with CppCodegen that I'm comfortable with (I would need to dig into factoring of `ComputeDependencyNodeDependencies`, make Compilation a generic type over the type of the node that represents methods and start sharing codepaths where we're going to do multithreaded compilation in RyuJIT).

I would prefer not to share those parts yet, or do it in a separate commit that is easy to revert when it gets inconvenient (because I have a feeling it will).